openquake/hazardlib/correlation.py

```diff
@@ -108,42 +136,199 @@ def __init__(self, vs30_clustering):
 
     def _get_correlation_matrix(self, sites, imt):
-        return jbcorrelation(sites, imt, self.vs30_clustering)
-
-    def get_lower_triangle_correlation_matrix(self, sites, imt):
         """
-        See :meth:`BaseCorrelationModel.get_lower_triangle_correlation_matrix`.
-        """
-        return numpy.linalg.cholesky(self._get_correlation_matrix(sites, imt))
+        Calculate correlation matrix for a given sites collection.
 
+        Correlation depends on spectral period, Vs 30 clustering behaviour
+        and distance between sites.
 
-def jbcorrelation(sites_or_distances, imt, vs30_clustering=False):
+        Parameters are the same as for
+        :meth:`BaseCorrelationModel.get_lower_triangle_correlation_matrix`.
         """
-        Returns the Jayaram-Baker correlation model.
+        distances = sites.mesh.get_distance_matrix()
+        return self._get_correlation_model(distances, imt)
 
-        :param sites_or_distances:
-            SiteCollection instance o ristance matrix
-        :param imt:
-            Intensity Measure Type (PGA or SA)
-        :param vs30_clustering:
-            flag, defalt false
+    def _get_correlation_model(self, distances, imt):
         """
-        if hasattr(sites_or_distances, 'mesh'):
-            distances = sites_or_distances.mesh.get_distance_matrix()
+        Returns the correlation model for a set of distances, given the
+        appropriate period
+
+        :param numpy.ndarray distances:
+            Distance matrix
+
+        :param float period:
+            Period of spectral acceleration
+        """
+        if isinstance(imt, SA):
+            period = imt.period
         else:
-            distances = sites_or_distances
+            assert isinstance(imt, PGA), imt
+            period = 0
 
         # formulae are from page 1700
-        if imt.period < 1:
-            if not vs30_clustering:
+        if period < 1:
+            if not self.vs30_clustering:
                 # case 1, eq. (17)
-                b = 8.5 + 17.2 * imt.period
+                b = 8.5 + 17.2 * period
             else:
                 # case 2, eq. (18)
-                b = 40.7 - 15.0 * imt.period
+                b = 40.7 - 15.0 * period
         else:
             # both cases, eq. (19)
-            b = 22.0 + 3.7 * imt.period
+            b = 22.0 + 3.7 * period
 
         # eq. (20)
         return numpy.exp((- 3.0 / b) * distances)
+
+    def get_lower_triangle_correlation_matrix(self, sites, imt, stddev_intra):
+        """
+        See :meth:`BaseCorrelationModel.get_lower_triangle_correlation_matrix`.
+        """
+        # Reshape 'stddev_intra' if needed.
+        stddev_intra_shape = stddev_intra.shape
+        try:
+            if stddev_intra_shape[0] > stddev_intra_shape[1]:
+                stddev_intra = numpy.transpose(stddev_intra)
+                stddev_intra = stddev_intra[0]
+        except:
+            stddev_intra = stddev_intra
```

Comment:
Please specify the expected Exception

Reply:
Same as before, I'll include the IndexError exception